lib/IRGen/GenConstant.cpp

```diff
@@ -0,0 +1,131 @@
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/InlineAsm.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallBitVector.h"
+#include "llvm/ADT/TinyPtrVector.h"
+
+#include "llvm/IR/Constant.h"
+#include "IRGenModule.h"
+#include "GenStruct.h"
+#include "TypeInfo.h"
+
+using namespace swift;
+using namespace irgen;
+
+/// Construct a ConstantInt from an IntegerLiteralInst.
+llvm::Constant *IRGenModule::emitConstantInt(swift::IntegerLiteralInst *ILI) {
+  APInt value = ILI->getValue();
+  BuiltinIntegerWidth width
+    = ILI->getType().castTo<BuiltinIntegerType>()->getWidth();
+
+  // The value may need truncation if its type had an abstract size.
+  if (width.isFixedWidth()) {
+    // nothing to do
+  } else if (width.isPointerWidth()) {
+    unsigned pointerWidth = getPointerSize().getValueInBits();
+    assert(pointerWidth <= value.getBitWidth()
+           && "lost precision at AST/SIL level?!");
+    if (pointerWidth < value.getBitWidth())
+      value = value.trunc(pointerWidth);
+  } else {
+    llvm_unreachable("impossible width value");
+  }
+
+  return llvm::ConstantInt::get(LLVMContext, value);
+}
+
+/// Construct a ConstantFP from a FloatLiteralInst.
+llvm::Constant *IRGenModule::emitConstantFP(swift::FloatLiteralInst *FLI) {
+  return llvm::ConstantFP::get(LLVMContext, FLI->getValue());
+}
+
+llvm::Constant *IRGenModule::emitAddrOfConstantString(StringLiteralInst *SLI) {
+  switch (SLI->getEncoding()) {
+ case swift::StringLiteralInst::Encoding::UTF8:
+   return getAddrOfGlobalString(SLI->getValue());
+
+ case swift::StringLiteralInst::Encoding::UTF16: {
+   // This is always a GEP of a GlobalVariable with a nul terminator.
+   auto addr = getAddrOfGlobalUTF16String(SLI->getValue());
+
+   // Cast to Builtin.RawPointer.
+   return llvm::ConstantExpr::getBitCast(addr, Int8PtrTy);
+ }
+
+ case swift::StringLiteralInst::Encoding::ObjCSelector:
+   llvm_unreachable("cannot get the address of an Objective-C selector");
+  }
+  llvm_unreachable("bad string encoding");
+}
+
+template<typename Inst>
+static llvm::Constant *emitConstantStructOrTuple(IRGenModule &IGM, Inst *SI) {
+  SmallVector<llvm::Constant*, 32> Elts;
+
+  auto Type = SI->getType();
+  auto PaddingPositionBits = irgen::getPhysicalStructPaddingPositionBits(IGM, Type);
+  auto *STy = cast<llvm::StructType>(IGM.getTypeInfo(Type).getStorageType());
+
+
+  auto PaddingFields = PaddingPositionBits.count();
+  auto NonPaddingFields = PaddingPositionBits.size() - PaddingFields;
+  assert(PaddingPositionBits.size() == STy->getNumElements() &&
+         "mismatch constant struct's padding layout and its storage type");
+  assert(NonPaddingFields == SI->getElements().size() &&
+         "mismatch constant struct's padding layout and its initialisers");
+
+  unsigned NonPaddingCount = 0;
+  for (unsigned i = 0, e = STy->getNumElements(); i != e; ++i) {
```

Comment:
A better algorithm here is to fill out the array with nulls, then walk the components and fill in the appropriate field with the initializer for that component, then fill in any remaining entries with padding items.

That would also allow you to avoid remembering padding indexes.